Prevent broker scale down (#54)
strimzi/proposals: prevent-broker-scale-down-if-contains-partition .md

> # Prevent scale down of the brokers if they contain replicas

Reviewer: as already done for the auto-rebalancing proposal, let's try to use `partition replicas` instead of just `replicas` everywhere.

Author: Sure on it

This proposal is about adding the mechanism for preventing the scale down of the brokers in the cluster if one or more replicas are hosted on it.

> In the future also, when we introduce the automatic re-balancing when scaling up/down the Kafka Cluster ([proposal](https://github.com/strimzi/proposals/blob/main/035-rebalance-types-scaling-brokers.md)), this mechanism would helpful to prevent any data loss when we are scaling down the brokers when they have some partitions assigned to them.

Reviewer: add the reference to the PR #57 and not the 035 you have here.

## Motivation

We should definitely introduce some sort of logic which can detect if the broker which is going to be removed still contains the partition replicas or not.
If the partition replicas are still available on the broker which is going to be scaled, then we should get some warning in the logs which will prompt users to do the reassignment and hopefully reduce the chance of data loss.

## Proposal

This proposal suggest how we can add the check to detect if the broker still contains any partition replicas and what to do if the broker scale down is not possible.

### Checking if the broker contains partition

#### Flow :

- The best way to detect if the broker contains any partition replicas is to query the topic metadata.
- We can create an Admin client instance for this purpose which can be used to get connect with the cluster and get us the topic details(topic name and topic description)
- We can later on query the topic's description to get the information regarding the `topic partition replicas`.
- Then we can use this information to check if the broker contains any partition replicas or not.

> - We should make sure that there is no partial scale down by checking the replicas on the brokers(to be scaled down) first and only do the whole scale down after we make sure that the brokers that are going to be removed aren't containing any partition replicas.

Reviewer 2: Again, don't describe what we can do or should do. Just say what we will do.

#### What to do if a broker contains partitions?

> - If a partition is detected over a broker then we should update the Kafka CR with the desired warnings

Reviewer 2: Make it clear that the warnings will go to the status.

- In case, if the admin client is not able to connect to the cluster, or we are not able to somehow get the topic details then in that case we should not do any scale down and wait for the next reconciliation to happen, but we will update the Kafka CR with the corresponding error message. 
  Once the user decides to either change the `spec.replicas` back to the previous count or reassign the partition replicas to other brokers, we can move the process accordingly.

### Error Handling

> If we somehow fail to connect to the Kafka Cluster or there are some other error which happens during the reconciliation, in that case also we will set the `scaleDown` variable to `false`, and provide a warning log on reason why the `scaleDown` was not done.

Reviewer 2: I think this deserves more details. When will the check be done? Which steps are affected by not scaling down? You refer to some `scaleDown` variable, but it is not clear what it is, where and how is it used.

# Prevent scale down of the brokers if they contain replicas

This proposal is about adding the mechanism for preventing the scale down of the brokers in the cluster if one or more replicas are hosted on it.

> In the future, when we introduce the automatic re-balancing when scaling up/down the Kafka Cluster ([proposal](https://github.com/strimzi/proposals/blob/main/035-rebalance-types-scaling-brokers.md)), this mechanism would help to prevent any data loss when we are scaling down the brokers when they have some partitions assigned to them.

Reviewer 3: It think this is a good feature to have weather we have auto-rebalancing or not. Warning a user of dataloss would be a good for robustness.

## Current situation

> Currently, when the number of Kafka replicas is decreased in the Kafka CR the StatefulSet will shut down the highest numbered brokers and there is no check to see if these brokers contain any topic-partition replicas.

Reviewer 2: You can probably write this more generic:

```suggestion
Currently, when removing brokers from the Kafka cluster, there is no check to see if these brokers contain any partition replicas.
```

> The [documentation](https://strimzi.io/docs/operators/latest/configuring.html#scaling-clusters-str) recommends that all topic-partition replicas are reassigned before doing this scaling action, which means that if someone tries to scale down without re-assigning the brokers, it can cause data loss.

Reviewer 2: ```suggestion
The [documentation](https://strimzi.io/docs/operators/latest/configuring.html#scaling-clusters-str) recommends that all topic-partition replicas are reassigned before doing this scaling action, which means that if someone tries to scale down without re-assigning the brokers, it can cause availability issues or data loss.
```

## Motivation

We should introduce the logic which can detect if the broker which is going to be removed still contains the partition replicas or not.

> If the partition replicas are still assigned to the broker which is going to be scaled up/down, then we should get some warning in the status of the Kafka resource which will prompt users to do the reassignment and hopefully reduce the chance of data loss.

Reviewer 2: ```suggestion
If any partition replicas are still assigned to the broker which is going to be removed from the cluster, then we should get some warning in the status of the Kafka resource which will prompt users to do the reassignment and prevent the broker from being removed until the partition replicas are reassigned.
```

## Proposal

This proposal suggest how we can add the check to detect if the broker still contains any partition replicas and what to do if the broker scale down is not possible.

### Checking if the broker contains partition

#### Flow :

> - The best way to detect if the broker contains any partition replicas is to query the topic metadata.

Reviewer 2: Again, write it as a fact:

```suggestion
- The topic metadata will be queried to detect if the broker contains any partition replicas.
```

> - We can create an Admin client instance for this purpose which can be used to get connect with the cluster and get us the topic details(topic name and topic description)

Reviewer 2: ```suggestion
- An Admin client instance will be used to get connect with the cluster and get us the topic details(topic name and topic description)
```

> - We can later on query the topic's description to get the information regarding the `topic partition replicas`.

Reviewer 2: How does this differ from the previous point?

- Then we can use this information to check if the broker contains any partition replicas or not.
- The scale down is done after we make sure that the brokers that are going to be removed aren't containing any partition replicas. By doing this we avoid any partial scale down.

#### What to do if a broker contains partitions?

#### Flow

> - A check will be placed just before the scale down steps happens. 

Reviewer 2: It probably needs to be placed earlier to make sure that everything is proerly configured if the scaledown needs to be aborted.

> - If a partition is detected over a broker then the Kafka CR will be updated with the desired warnings.

Reviewer 2: This is a bit confusing ... you should make it clear in this sentence if it is just the sattus of the Kafka CR which will be updated.

  The warnings will be displayed in the status.

> - When the scale down is happening, if the admin client is not able to connect to the cluster, or not able to get the topic details, scale down will not be proceeded, and we will wait for the next reconciliation to happen.

Reviewer 2: This should have more details about how it will be done. You also need to explain what will you do when it connects but the partitions are still assigned to the removed brokers.

Once the user decides to either change the `spec.replicas` back to the previous count or reassign the partition replicas to other brokers, we can move the process accordingly.
  

> 

Reviewer 2: No need for 3 empty lines I guess.

## Current situation

Currently, when the number of Kafka replicas is decreased in the Kafka CR the StatefulSet will shut down the highest numbered brokers and there is no check to see if these brokers contain any topic-partition replicas.
The [documentation](https://strimzi.io/docs/operators/latest/configuring.html#scaling-clusters-str) recommends that all topic-partition replicas are reassigned before doing this scaling action, which means that if someone tries to scale down without re-assigning the brokers, it can cause data loss.

## Motivation

We should definitely introduce some sort of logic which can detect if the broker which is going to be removed still contains the partitions or not.
If the partitions are still available on the broker which is going to be scaled, then we should get some warning in the logs which will prompt users to do the reassignment and hopefully reduce the chance of data loss.

## Proposal

This proposal suggest how we can add the check to detect if the broker still contains any partitions and what to do if the broker scale down is not possible.

> Checking if the broker contains partition

Reviewer 3: Is this a title?

> The best way to detect if the broker contains any partitions is to query the topics.

Reviewer 3: "is to query the cluster metadata"

To do this, we will have to make use of the `AdminClient` class.
We should probably put the logic for this check inside the `scaleDown` method present in the `KafkaReconciler` class.
The `scaleDown` method gets triggered when we scale down the no. of replicas in the `Kafka CR`.
So this check will target the broker which are going to be removed i.e. the pods with the highest number.

Methods to be created in the process:

- `getTopicNamesAndDescriptions` (Boolean Type) - This method will be placed inside the `scaleDown()` method in `KafkaReconciler` class and will provide us with the Collection of `TopicDescription`.
- `describeTopics()` and `topicNames()`  - These two methods will provide us with the description and names of all the topics ie first we will get the names of the topics using the `topicNames` method and using these names we will extract the description of the topics.

> - `brokerHasAnyTopics()` - This method will take as input a podId and a `Collection<TopicDescription>`.

Reviewer 3: Would it not make more sense to do one loop over the TopicDescription collection and produce a simple map from node ID to a boolean. That way if (in future) you want to scale down more than one broker at a time you won't have to re-loop over the collection you can just use the `hasAssignedTopicPartitions` map?

  It will go through all the broker and see if any partitions are assigned to it.

### Flow :

When the replicas are changed inside the Kafka resource, reconciliation will happen.
During this reconciliation, we move to the `scaleDown()` method present in the `KafkaReconciler` class.
The `scaleDown()` method contains the complete mechanism on how the broker is scaled down. So there, we will first create our `adminClient` and provide it with the Secrets.
Then we will call the `getTopicNamesAndDescriptions()` method which will give us the `Collection<TopicDescription>`. 

The `getTopicNamesAndDescriptions()` method will call the `topicNames()` method to get the set of topic names and then these topic names will be used by the `describeTopics()` methods to get the `Collection<TopicDescription>`.

> Once the `Collection<TopicDescription>` is computed we will move back to the `scaledown()` method and there we will run a loop from `(previous replicas - 1)` to `current replicas` and inside the loop we will call the `brokersHasAnyTopics()` method which will take the `Collection<TopicDescription>` and `podId(index of the loop)` to check if the broker contains any partition and provide us with a boolean result. 

Reviewer 3: See comment above, I think its better to just loop over the collection once and produce a map from nodeid to boolean.

### What to do if a broker contains partitions?
If a partition is detected over a broker then we should trigger some log warnings and let the partitions remain the same as they were before.

> For example, if we try to scale down the broker from 5 to 3, but the check detected that the broker to be scaled down contains a partition then we should keep the replicas as 5 only display the appropriate warnings.

Reviewer 3: "the the brokers to be scaled down still contain partitions"

> To handle the case, if broker contain some partitions, we can make use of an `if` statement. If we see the current code in the `scaleDown()` method we can see that we are checking whether we are using `StatefulSets` or `StrimziPodSets` and then doing the scale down accordingly, so we can introduce an `if` just before this logic. This `if` condition will check if the boolean result was `true` or `false`. If the result is `false` then we will return a succeeded future i.e. no need to scale down but if the result is `true` then we can run the logic of scale down as per `StatefulSets` or `StrimziPodSets`.

Reviewer 3: I don't understand this paragraph. If what result is false? What is this if statement checking? Weather we are using STSs or SPSs? Or if the broker has partitions?

Author: The `if` condition will check the boolean result stored in a variable w.r.t to whether the broker had partitions or not. If the result comes out to be `True` we will just return a succeeded future and skip the other logic `STS` and `SPS` since we don't need to scale down at all

Reviewer: I think this comment is related to my more general comment where I am saying that the proposal sounds to describe too many programming constructs than describing the overall solution. I found this paragraph hard to understand too.

Reviewer 3: Yeah the proposal should stick to high level processes and discriptions and not talk about implementation.

The best way to detect if the broker contains any partitions is to query the topics.
To do this, we will have to make use of the `AdminClient` class.
We should probably put the logic for this check inside the `scaleDown` method present in the `KafkaReconciler` class.
The `scaleDown` method gets triggered when we scale down the no. of replicas in the `Kafka CR`.
So this check will target the broker which are going to be removed ie. the pods with the highest number.

Methods to be created in the process:

- `getTopicNameAndDescription` (Boolean Type) - This method will be placed inside the `scaleDown()` method in `KafkaReconciler` class and will provide us with the Collection of `TopicDescription`.
- `describeTopics()` and `topicNames()`  - These two methods will provide us with the description and names of all the topics ie first we will get the names of the topics using the `topicNames` method and using these names we will extract the description of the topics.
- `brokerHasAnyTopics()` - This method will take as input a podId and a `Collection<TopicDescription>`.
  It will go through all the broker and see if any partitions are assigned to it.

> ### Flow :

Reviewer 2: No space before `:`.

When the replicas are changed inside the Kafka resource, reconciliation will happen.
During this reconciliation, we move to the `scaleDown()` method present in the `KafkaReconciler` class.
The `scaleDown()` method contains the complete mechanism on how the broker is scaled down. So there, we will first create our `adminClient` and provide it with the Secrets.

> Then we will call the `checkPartitions()` method which will give us the `Collection<TopicDescription>`. 

Reviewer 2: You should elaborate on how the error handling will look like here when it fails to connect to the Kafka cluster.

Author: @scholzj Hi, do you mean what if any `RuntimeException` pops up and then how will I handle the process to not stop the reconciliation process?

> The `checkPartition()` method will call the `topicNames()` method to get the set of topic names and then these topic names will be used by the `describeTopics()` methods to get the `Collection<TopicDescription>`.

Reviewer 2: `checkPartitions()`?

> Once the `Collection<TopicDescription>` is computed we will move back to the `scaledown()` method and there we will run a loop from `(previous replicas - 1)` to `current replicas` and inside the loop we will call the `brokersHasAnyTopics()` method which will take the `Collection<TopicDescription>` and `podId(index of the loop)` to check if the broker contains any partition and provide us with a boolean result.

Reviewer 2: This sounds like a hard to handle situation where you do the scale down _partially_. It might be easier to do none or all.

### What to do if a broker contains partitions?
If a partition is detected over a broker then we should trigger some log warnings and let the partitions remain the same as they were before.

> For example, if we try to scale down the broker from 5 to 3, but the check detected that the broker to be scaled down contains a partition then we should keep the replicas as 5 only display the appropriate warnings.

Reviewer 2: This needs more details on how do you achieve keeping 5 replicas. It sounds easier than it might be.

> # Prevent scale down of the brokers if they contain partitions

Reviewer: I would talk in general about replicas not partition. They could be the same if the topic has 1 partition with 1 replica but not when it has more replicas. We should not scale down in any case, if the broker is hosting a replica leader or a follower (to avoid under-replicated partition or less than min ISR).

This proposal is about adding the mechanism for preventing the scale down of the brokers in the cluster if a partition is assigned to it.

> In the future, when we introduce the automatic scale up/down of the Kafka Cluster, this mechanism would help to prevent any data loss when we are scaling down the brokers when they have some partitions assigned to them.

Reviewer: Which "automatic scale up/down" are you referring to?

Author: Oh my bad, Let me address this

Author: I was actually trying to point towards the next steps of the Automatic add/ remove of the brokers from cluster. Like cureently we have introduced modes in the rebalance but in future we were going to do all the process without user interaction

Reviewer: so more than "automatic scale up/down of the Kafka Cluster" is "automatic rebalancing when scaling up/down the Kafka cluster". I would also link to the [proposal](https://github.com/strimzi/proposals/blob/main/035-rebalance-types-scaling-brokers.md) where we mention this future plan, so that people know what is in place.

## Current situation

> Currently, when the number of Kafka replicas is decreased in the Kafka CR the StatefulSet will shut down the highest numbered brokers and there is no check to see if these brokers contain any topic-partition replicas.

Reviewer: you are doing better here mentioning "topic-partition replicas" :-)

The [documentation](https://strimzi.io/docs/operators/latest/configuring.html#scaling-clusters-str) recommends that all topic-partition replicas are reassigned before doing this scaling action, which means that if someone tries to scale down without re-assigning the brokers, it can cause data loss.

## Motivation

We should definitely introduce some sort of logic which can detect if the broker which is going to be removed still contains the partitions or not.

> If the partitions are still available on the broker which is going to be scaled, then we should get some warning in the logs which will prompt users to do the reassignment and hopefully reduce the chance of data loss.

Reviewer: People not having Cruise Control enabled will have only manual reassignment as the way to go.
I would mention that for people having Cruise Control, they could run a rebalancing in `remove-brokers` mode.
Refer to the corresponding proposal, and maybe documentation if it will be merged before this proposal got accepted.

Reviewer 4: > we should get some warning in the logs

You cannot assume that anyone is reading the logs, so a warning achieves very little. I think we should fail the reconciliation in this case.

Author: @tombentley Hi, I thought we don't have to fail a reconciliation if somehow there were replicas present on broker. As per the discussion with Jakub, I thought we had to let the reconciliation take place without any scale down of brokers but the reconciliation should still happen

Reviewer: Failing the reconciliation would mean anyway updating the Kafka CR with a warning condition about the error, so the user knows why it is not scaling.

Reviewer 2: You cannot fail the reconciliation. The cluster might be in a bad state, but because the scaledown is fairly early in the reconciliation, it would cause the operator to never fix it and in turn the scale-down will never succeed. So it needs to pass:
* When there is any problem with connecting to the broker(s)
* When the partitions don't allow scale-down (because other brokers might have issues and not allow the partitions to move)

## Proposal

This proposal suggest how we can add the check to detect if the broker still contains any partitions and what to do if the broker scale down is not possible.

### Checking if the broker contains partition
The best way to detect if the broker contains any partitions is to query the topic metadata.
To do this, we will have to make use of the `AdminClient` class.

> We should probably put the logic for this check inside the `scaleDown` method present in the `KafkaReconciler` class.

Reviewer 4: "Probably"? In a proposal you should be clear about exactly what you're proposing. 
```suggestion
We will put the logic for this check inside the `scaleDown` method present in the `KafkaReconciler` class.
```

The `scaleDown` method gets triggered when we scale down the no. of replicas in the `Kafka CR`.
So this check will target the broker which are going to be removed i.e. the pods with the highest number.

Methods to be created in the process:

- `getTopicNamesAndDescriptions` (Boolean Type) - This method will be placed inside the `scaleDown()` method in `KafkaReconciler` class and will provide us with the Collection of `TopicDescription`.
- `describeTopics()` and `topicNames()`  - These two methods will provide us with the description and names of all the topics ie first we will get the names of the topics using the `topicNames` method and using these names we will extract the description of the topics.

> - `brokerContainsPartitions` - This method will take as input a `Collection<TopicDescription>` and then generate a map of type `HashMap<Integer podId, Boolean containsPartitions>`.

Reviewer 4: Personally I don't think this level of detail is needed in a proposal. It's an implementation detail.

### Flow :

> When the replicas are changed inside the Kafka resource, reconciliation will happen.

Reviewer 4: ```suggestion
When the `spec.kafka.replicas` are changed inside the Kafka resource, reconciliation will happen.
```

During this reconciliation, we move to the `scaleDown()` method present in the `KafkaReconciler` class.
The `scaleDown()` method contains the complete mechanism on how the broker is scaled down. So there, we will first create our `adminClient` and provide it with the Secrets.
Then we will call the `getTopicNamesAndDescriptions()` method which will give us the `Collection<TopicDescription>`. 

The `getTopicNamesAndDescriptions()` method will call the `topicNames()` method to get the set of topic names and then these topic names will be used by the `describeTopics()` methods to get the `Collection<TopicDescription>`. 
Once the `Collection<TopicDescription>` is computed we will move back to the `scaledown()` method and there we can call `brokerContainsPartition()` method which will take the `Collection<TopicDescription>` and then create a `hasAssignedTopicPartitions` map which contains key as `podId` and value as `boolean` type.
Now we can use this map to check if the broker contains any partitions or not.

### What to do if a broker contains partitions?

> If a partition is detected over a broker then we should trigger some log warnings and let the partitions remain the same as they were before.

Reviewer 4: Again, fail the reconciliation, since logging won't actually prevent data loss.

Reviewer: But updating the Kafka CR with a warning condition.
